src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java

```diff
@@ -0,0 +1,115 @@
+        return new NumberConfigType<>(
+                new DecimalSerializableType(minValue == null ? null : f0.apply(minValue), maxValue == null ? null : f0.apply(maxValue), precision == null ? null : f0.apply(precision)),
+                actualType,
+                f,
+                f0
+        );
+    }
+
+    /* String-derived types */
+
+    public static final StringConfigType<String> STRING =
+            new StringConfigType<>(StringSerializableType.DEFAULT_STRING, String.class, Function.identity(), Function.identity());
+
+    public static final StringConfigType<Character> CHARACTER =
+            STRING.withMinLength(1).withMaxLength(1).derive(Character.class, s -> s.charAt(0), Object::toString);
+
+    /* Enum-derived types */
+
+    public static <E extends Enum<E>> EnumConfigType<E> makeEnum(Class<E> enumType) {
+        Set<String> validValues = Arrays.stream(enumType.getEnumConstants()).map(Enum::name).collect(Collectors.toSet());
+        return new EnumConfigType<>(new EnumSerializableType(validValues), enumType, e -> Enum.valueOf(enumType, e), Enum::name);
+    }
+
+    /* List-derived types */
+
+    public static <E0, E, U extends ConfigType<E, E0, ?>> ListConfigType<List<E>, E0> makeList(U elementType) {
+        return new ListConfigType<>(
+                ListSerializableType.of(elementType.getSerializedType()), List.class,
+                l -> {
+                    List<E> ret = new ArrayList<>();
+                    for (E0 e0 : l) {
+                        ret.add(elementType.toRuntimeType(e0));
+                    }
+                    return Collections.unmodifiableList(ret);
+                },
+                l -> {
+                    List<E0> ret = new ArrayList<>();
+                    for (E e : l) {
+                        ret.add(elementType.toSerializedType(e));
+                    }
+                    return Collections.unmodifiableList(ret);
+                }
+        );
+    }
+
+    public static <E0, E, U extends ConfigType<E, E0, ?>> ListConfigType<Set<E>, E0> makeSet(U elementType) {
+        return makeList(elementType).withUniqueElements().derive(Set.class, HashSet::new, ArrayList::new);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <E0, E, U extends ConfigType<E, E0, ? extends SerializableType<E0>>> ListConfigType<E[], E0> makeArray(U elementType) {
+        E[] z = (E[]) Array.newInstance(elementType.getRuntimeType(), 0);
+        Class<E[]> arrType = (Class<E[]>) z.getClass();
+        // account for arrays of primitives - we must avoid autocasting to Object[]
+        return (ListConfigType<E[], E0>) (ListConfigType<?, ?>) makeList((ConfigType<Object, Object, ?>) elementType).derive((Class<? super Object>) (Class<?>) arrType, l -> TypeMagic.unbox(l.toArray((Object[]) z)), a -> Arrays.asList((E[]) TypeMagic.box(a)));
+    }
```

Comment:
What about `ListConfigType<int[], Integer>`? If you want primitives, you may have to write a specialized method for each primitive type.

Reply:
This was absolutely ridiculous, ditched it for something vaguely reasonable.